scylla-cql/src/types/deserialize/result.rs

```diff
@@ -0,0 +1,195 @@
+use crate::frame::response::result::ColumnSpec;
+
+use super::row::{mk_deser_err, BuiltinDeserializationErrorKind, ColumnIterator, DeserializeRow};
+use super::{DeserializationError, FrameSlice, TypeCheckError};
+use std::marker::PhantomData;
+
+/// Iterates over the whole result, returning rows.
+pub struct RowIterator<'frame> {
+    specs: &'frame [ColumnSpec],
+    remaining: usize,
+    slice: FrameSlice<'frame>,
+}
+
+impl<'frame> RowIterator<'frame> {
+    /// Creates a new iterator over rows from a serialized response.
+    ///
+    /// - `remaining` - number of the remaining rows in the serialized response,
+    /// - `specs` - information about columns of the serialized response,
+    /// - `slice` - a `FrameSlice` that points to the serialized rows data.
+    #[inline]
+    pub fn new(remaining: usize, specs: &'frame [ColumnSpec], slice: FrameSlice<'frame>) -> Self {
+        Self {
+            specs,
+            remaining,
+            slice,
+        }
+    }
+
+    /// Returns information about the columns of rows that are iterated over.
+    #[inline]
+    pub fn specs(&self) -> &'frame [ColumnSpec] {
+        self.specs
+    }
+
+    /// Returns the remaining number of rows that this iterator is supposed
+    /// to return.
+    #[inline]
+    pub fn rows_remaining(&self) -> usize {
+        self.remaining
+    }
+}
+
+impl<'frame> Iterator for RowIterator<'frame> {
+    type Item = Result<ColumnIterator<'frame>, DeserializationError>;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        self.remaining = self.remaining.checked_sub(1)?;
+
+        let iter = ColumnIterator::new(self.specs, self.slice);
+
+        // Skip the row here, manually
+        for (column_index, spec) in self.specs.iter().enumerate() {
+            if let Err(err) = self.slice.read_cql_bytes() {
+                return Some(Err(mk_deser_err::<Self>(
+                    BuiltinDeserializationErrorKind::RawColumnDeserializationFailed {
+                        column_index,
+                        column_name: spec.name.clone(),
+                        err: DeserializationError::new(err),
+                    },
+                )));
+            }
+        }
+
+        Some(Ok(iter))
+    }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (0, Some(self.remaining))
```

Comment:
This 0 should be self.remaining. Iterator only returns `None` after `self.remaining` becomes 0, which happends precisely after `self.remaining` iterations